File: src/core/contracts/bid-action.ts
/**
 * Canonical Observation Ontology — bridge-universal semantic observations.
 *
 * Each `BidAction` describes a single bridge communicative act in convention-erased
 * terms. Modules produce convention-shaped `sourceIntent` strings ("StaymanAsk",
 * "DONTRevealClubs"); the normalization layer translates them into canonical
 * observations ("inquire(majorSuit)", "show(heldSuit, clubs)").
 *
 * **Stability goal:** The 17 acts below cover all current conventions plus forward
 * provisions for RKCB, Control Cuebids, NT stoppers, and competitive bidding. When
 * authoring a new convention, use existing acts with new parameterizations first.
 * Only propose a new act when no existing act captures the communicative semantics.
 * Expect at most 1–2 additions for slam or advanced competitive conventions.
 *
 * **Consumers:** Pattern matchers (continuation composition), CommittedStep pipeline,
 * strategy layer. Producers: `normalizeIntent()` in `conventions/core/pipeline/`.
 */

// ── Parameter domains ──────────────────────────────────────────────

export type ObsSuit = "clubs" | "diamonds" | "hearts" | "spades";
export type BidSuitName = ObsSuit | "notrump";

export type HandFeature =
  | "heldSuit" // holding a specific suit (5+ cards or 4 if shown)
  | "shortMajor" // 4-card major (Smolen: bid suit = short major)
  | "majorSuit" // 4+ card major (unspecified which)
  | "twoSuited" // two-suited hand
  | "suitQuality" // suit texture (good/bad/solid)
  | "strength" // hand strength generally
  | "shortage" // short suit (void/singleton/doubleton)
  | "fit" // fit for partner's suit
  | "balanced" // balanced shape
  | "control" // control in a suit (ace/void/king/singleton)
  | "keyCards" // RKCB keycards
  | "stopper"; // suit stopper for NT

export type HandStrength =
  | "minimum"
  | "maximum"
  | "invitational"
  | "game"
  | "slam"
  | "preemptive"
  | "constructive"
  | "limit"
  | "weak"
  | "strong";

export type SuitQuality = "good" | "bad" | "solid";

// ── Observation acts ───────────────────────────────────────────────

/**
 * Discriminated union of bridge communicative acts.
 *
 * Each variant models a single semantic observation. One `sourceIntent` may
 * produce multiple observations (e.g., Splinter = shortage + raise).
 *
 * **`overcall` vs `show`:** `overcall` marks competitive entry — structurally
 * distinct from cooperative `show`. A natural 2H overcall is
 * `overcall(heldSuit, hearts)`. A 2H response to partner is
 * `show(heldSuit, hearts)`. Position (competitive vs cooperative) matters
 * for inference.
 */
export type BidAction =
  | { readonly act: "open"; readonly strain: BidSuitName; readonly strength?: HandStrength }
  | {
      readonly act: "show";
      readonly feature: HandFeature;
      readonly suit?: ObsSuit;
      readonly quality?: SuitQuality;
      readonly strength?: HandStrength;
    }
  | { readonly act: "deny"; readonly feature: HandFeature; readonly suit?: ObsSuit }
  | { readonly act: "inquire"; readonly feature: HandFeature; readonly suit?: ObsSuit }
  | { readonly act: "transfer"; readonly targetSuit: ObsSuit }
  | {
      readonly act: "accept";
      readonly feature: HandFeature;
      readonly suit?: ObsSuit;
      readonly strength?: HandStrength;
    }
  | { readonly act: "decline"; readonly feature: HandFeature; readonly suit?: ObsSuit }
  | { readonly act: "raise"; readonly strain: BidSuitName; readonly strength: HandStrength }
  | { readonly act: "place"; readonly strain: BidSuitName }
  | { readonly act: "signoff"; readonly strain?: BidSuitName }
  | { readonly act: "force"; readonly level: HandStrength }
  | { readonly act: "agree"; readonly strain: BidSuitName }
  | { readonly act: "relay"; readonly forced: boolean }
  | { readonly act: "overcall"; readonly feature: HandFeature; readonly suit?: ObsSuit }
  | { readonly act: "double"; readonly feature: HandFeature }
  | { readonly act: "pass" }
  | { readonly act: "redouble"; readonly feature: HandFeature };

// ── Runtime enumeration ────────────────────────────────────────────

/** All 17 act type strings — for exhaustive matching and iteration. */
export const BID_ACTION_TYPES = [
  "open",
  "show",
  "deny",
  "inquire",
  "transfer",
  "accept",
  "decline",
  "raise",
  "place",
  "signoff",
  "force",
  "agree",
  "relay",
  "overcall",
  "double",
  "pass",
  "redouble",
] as const;

export type BidActionType = (typeof BID_ACTION_TYPES)[number];

File: src-tauri/crates/bridge-conventions/src/types/bid_action.rs
//! Canonical observation ontology — bridge-universal semantic observations.
//!
//! Mirrors TS types from `conventions/pipeline/bid-action.ts`.

use serde::{Deserialize, Serialize};

/// Observation suit (long-form names, not engine single-char).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ObsSuit {
    Clubs,
    Diamonds,
    Hearts,
    Spades,
}

/// Suit name including notrump (used across convention types).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum BidSuitName {
    Clubs,
    Diamonds,
    Hearts,
    Spades,
    Notrump,
}

/// Hand feature being communicated by a bid action.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum HandFeature {
    HeldSuit,
    ShortMajor,
    MajorSuit,
    TwoSuited,
    SuitQuality,
    Strength,
    Shortage,
    Fit,
    Balanced,
    Control,
    KeyCards,
    Stopper,
}

/// Hand strength level.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum HandStrength {
    Minimum,
    Maximum,
    Invitational,
    Game,
    Slam,
    #[serde(rename = "slam-invite")]
    SlamInvite,
    Preemptive,
    Constructive,
    Limit,
    Weak,
    Strong,
}

/// Suit quality descriptor.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SuitQuality {
    Good,
    Bad,
    Solid,
}

/// All known bid action type strings.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum BidActionType {
    Open,
    Show,
    Deny,
    Inquire,
    Transfer,
    Accept,
    Decline,
    Raise,
    Place,
    Signoff,
    Force,
    Agree,
    Relay,
    Overcall,
    Double,
    Pass,
    Redouble,
}

/// Discriminated union of bridge communicative acts.
///
/// Tagged on the `act` field to match TS serialization.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "act")]
pub enum BidAction {
    #[serde(rename = "open")]
    Open {
        strain: BidSuitName,
        #[serde(skip_serializing_if = "Option::is_none")]
        strength: Option<HandStrength>,
    },
    #[serde(rename = "show")]
    Show {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
        #[serde(skip_serializing_if = "Option::is_none")]
        quality: Option<SuitQuality>,
        #[serde(skip_serializing_if = "Option::is_none")]
        strength: Option<HandStrength>,
    },
    #[serde(rename = "deny")]
    Deny {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
    },
    #[serde(rename = "inquire")]
    Inquire {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
    },
    #[serde(rename = "transfer")]
    Transfer { #[serde(rename = "targetSuit")] target_suit: ObsSuit },
    #[serde(rename = "accept")]
    Accept {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
        #[serde(skip_serializing_if = "Option::is_none")]
        strength: Option<HandStrength>,
    },
    #[serde(rename = "decline")]
    Decline {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
    },
    #[serde(rename = "raise")]
    Raise {
        strain: BidSuitName,
        strength: HandStrength,
    },
    #[serde(rename = "place")]
    Place { strain: BidSuitName },
    #[serde(rename = "signoff")]
    Signoff {
        #[serde(skip_serializing_if = "Option::is_none")]
        strain: Option<BidSuitName>,
    },
    #[serde(rename = "force")]
    Force { level: HandStrength },
    #[serde(rename = "agree")]
    Agree { strain: BidSuitName },
    #[serde(rename = "relay")]
    Relay { forced: bool },
    #[serde(rename = "overcall")]
    Overcall {
        feature: HandFeature,
        #[serde(skip_serializing_if = "Option::is_none")]
        suit: Option<ObsSuit>,
    },
    #[serde(rename = "double")]
    Double { feature: HandFeature },
    #[serde(rename = "pass")]
    Pass,
    #[serde(rename = "redouble")]
    Redouble { feature: HandFeature },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bid_action_open_roundtrip() {
        let action = BidAction::Open {
            strain: BidSuitName::Notrump,
            strength: None,
        };
        let json = serde_json::to_string(&action).unwrap();
        assert!(json.contains(r#""act":"open""#));
        let back: BidAction = serde_json::from_str(&json).unwrap();
        assert_eq!(back, action);
    }

    #[test]
    fn bid_action_show_with_all_fields() {
        let action = BidAction::Show {
            feature: HandFeature::HeldSuit,
            suit: Some(ObsSuit::Hearts),
            quality: Some(SuitQuality::Good),
            strength: Some(HandStrength::Invitational),
        };
        let json = serde_json::to_string(&action).unwrap();
        let back: BidAction = serde_json::from_str(&json).unwrap();
        assert_eq!(back, action);
    }

    #[test]
    fn bid_action_pass_roundtrip() {
        let action = BidAction::Pass;
        let json = serde_json::to_string(&action).unwrap();
        assert_eq!(json, r#"{"act":"pass"}"#);
        let back: BidAction = serde_json::from_str(&json).unwrap();
        assert_eq!(back, action);
    }

    #[test]
    fn bid_action_transfer_roundtrip() {
        let action = BidAction::Transfer {
            target_suit: ObsSuit::Spades,
        };
        let json = serde_json::to_string(&action).unwrap();
        assert!(json.contains(r#""targetSuit":"spades""#));
        let back: BidAction = serde_json::from_str(&json).unwrap();
        assert_eq!(back, action);
    }

    #[test]
    fn hand_strength_serde() {
        assert_eq!(
            serde_json::to_string(&HandStrength::SlamInvite).unwrap(),
            "\"slam-invite\""
        );
        let back: HandStrength = serde_json::from_str("\"slam-invite\"").unwrap();
        assert_eq!(back, HandStrength::SlamInvite);
    }

    #[test]
    fn bid_action_type_serde() {
        assert_eq!(
            serde_json::to_string(&BidActionType::Open).unwrap(),
            "\"open\""
        );
        assert_eq!(
            serde_json::to_string(&BidActionType::Signoff).unwrap(),
            "\"signoff\""
        );
    }
}
